### PR TITLE
DBZ-713 Only getting collections for databases matching filters

### PR DESCRIPTION
**Pull request by gunnarmorling**
https://issues.jboss.org/browse/DBZ-713
https://issues.jboss.org/browse/DBZ-782

@jpechane This fixes two issues, I noticed DBZ-782 while working on DBZ-713. It's best reviewed commit by commit, as they are largely independent and I did some clean-up not strictly related to the original issue. Also disable whitelist differences to further ease the diff. Thanks!

In terms of testing, it seems all the MongoDB ITs we have don't deal with authentication/authorization at all. It seems worth to have some tests for that, but let's handle it separately. I tested the original issue manually successfully, i.e. collections from a DB the connector user isn't authorized for won't cause any harm if that other DB is not part of the filter config. I've added some basic tests which make sure that the methods of `MongoPrimary` only return the databases/collections matching the given filter config.

### CODE REVIEW

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Replicator.java [hunk 1]
```diff
@@ -169,9 +169,18 @@ public void run() {
...
     protected boolean establishConnectionToPrimary() {
         logger.info("Connecting to '{}'", replicaSet);
-        primaryClient = context.getConnectionContext().primaryFor(replicaSet, (desc, error) -> {
-            logger.error("Error while attempting to {}: {}", desc, error.getMessage(), error);
-        });
+        primaryClient = context.getConnectionContext().primaryFor(
+                replicaSet,
+                (desc, error) -> {
+                    // propagate authorization failures
+                    if (error.getMessage() != null && error.getMessage().startsWith("Command failed with error 13")) {
```
Comment (jpechane) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/Replicator.java]: This should be extracted as a constant

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/ConnectionContext.java [hunk 2]
```diff
@@ -281,6 +282,33 @@ public void execute(String desc, Consumer<MongoClient> operation) {
...
+         * Execute the supplied operation using the primary, blocking until a primary is available. Whenever the operation stops
+         * (e.g., if the primary is no longer primary), then restart the operation using the current primary.
+         *
+         * @param desc the description of the operation, for logging purposes
+         * @param operation the operation to be performed on the primary
+         * @return return value of the executed operation
+         */
+        public <T> T execute(String desc, Function<MongoClient, T> operation) {
+            final Metronome errorMetronome = Metronome.sleeper(PAUSE_AFTER_ERROR, Clock.SYSTEM);
+            while (true) {
```
Comment (jpechane) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/ConnectionContext.java]: Will this loop terminate when the connector is stopped?

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/ConnectionContext.java [hunk 3]
```diff
@@ -311,18 +339,19 @@ public void executeBlocking(String desc, BlockingConsumer<MongoClient> operation
          * @return the database names; never null but possibly empty
          */
         public Set<String> databaseNames() {
-            Set<String> databaseNames = new HashSet<>();
-            execute("get database names", primary -> {
-                databaseNames.clear(); // in case we restarted
```
Comment (jpechane) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/ConnectionContext.java]: Are you sure this line should be removed?

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/ConnectionContext.java [hunk 4]
```diff
@@ -334,10 +363,10 @@ public void executeBlocking(String desc, BlockingConsumer<MongoClient> operation
          */
         public List<CollectionId> collections() {
             String replicaSetName = replicaSet.replicaSetName();
+
             // For each database, get the list of collections ...
-            List<CollectionId> collections = new ArrayList<>();
-            execute("get collections in databases", primary -> {
-                collections.clear(); // in case we restarted
```
Comment (jpechane) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/ConnectionContext.java]: Same as above

**Comment by jpechane** [comment]
@gunnarmorling Good for me, just two questions were raise

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/ConnectionContext.java (hunk 2 above)
Comment (gunnarmorling) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/ConnectionContext.java]: Good point; indeed it seems they'd keep running if the connector is shut down and they don't complete normally. I'll log an issue.

**Comment by gunnarmorling** [comment]
@jpechane Pushed one more commit to extract the constant and filed [DBZ-784](https://issues.jboss.org/browse/DBZ-784) for the loop abortion issue.

**Comment by jpechane** [comment]
@gunnarmorling Applied, thanks.

**Pull request merged**